NPC Refactor

royjpark opened the pull request:
UPDATE  1:
- added new methods into CollisionManager, CollisionDetection, and Room to remove the npc from the room itself when they die because the removeCollder method within manager just would not work for some reason with the npcs. It was probably just me not knowing what to do, but I got it to work this way.
- implemented RNG movement for skeletons so far as a trial run, I'm liking it.
- fixed my issue with hitboxes mentioned above (dumb mistake on my part, didn't think to make the npc's hitboxes smaller as I kept trying to make them bigger)

UPDATE 2:
- added RNG movement to all other NPCs except boss and old man as those don't need it.
- reduced knockback distance so there's less of a chance for them to get stuck in wall.

UPDATE 3:
- enemy (just skeleton for now) knocks back in direction that Link is facing
- should be ready to merge after implementing ^ in Goriya

UPDATE 4:
- goriya knockback finished
- still some minor issues with obstacle collision, but that can be done in a separate PR.
- this should be able to be merged

LAST UPDATE (hopefully):
- fixed goriya boomerang issue
- fixed hitboxes and collision responses so no more getting stuck in walls
- added transparent doorblocks to prevent npcs from getting in open doors
- reverted bats from rng to movement cycle again (still torn on this one, but whichever one I choose it's easy to make the changes for the other classes)
- skeleton obstacle collision is a bit jerky but passable, hard not to be especially in the water map where it's surrounded by colliders.
- I believe that I've fixed everything to the best of my knowledge. I'll continue to work on the jerky skeleton but that can be another pr as it's very minor and since this should get merged asap.

royjpark committed:
fixed npc obstacle collision

royjpark committed:
Hitboxes work for left side and top side of enemy now, right and bottom are still jank

royjpark committed:
correct values for link facing east and facing south when using sword

royjpark committed:
?

royjpark committed:
Hitbox fully removed when NPC dies

royjpark committed:
implemented RNG movement for skeleton

bhasseman commented:
Everything you've done looks good so far, hitboxes seem much better also. Well done with that.
One thing that still needs fixed is when they are knocked back and their collision gets wacky with the walls/obstacles as they can still clip through them and either get stuck or get reset to a different position.

royjpark committed:
Implemented RNG to other classes, changed knockback distane

royjpark committed:
enemy now knocks back in direction that link is facing

royjpark committed:
Link direction knockback implemented in Goriya

rustbeltdurt commented:
Just took a look at this and ran it on my machine with no problems. Aside from a very few instances of wonky behavior, I think the hitboxes are passable now (as far as interactions with link's swords go). The A.I. pathfinding isn't really a concern for me, personally. The biggest things are that the NPCs can damage and be damaged. With that being said though, I think the NPC hitboxes could use a slight reduction (see how Link's hitbox was reduced), because the water rooms are not currently passable.

EDIT: Upon further investigation, the Boss also has a few discrepancies:
         1. It does not show any kind of damage (This doesn't need to be robust, but some basic feedback to show the user that the 
                hit was successful is necessary)
         2. Its movement is extremely slow and could probably use the RNG treatment that the other NPCs got.

EDIT2: Sorry for committing here, had to get this task in while I could. Should be easier to merge everything than if I took out a separate branch.

rustbeltdurt committed:
World Items spawn only when all enemies in the room are eliminated, fixed persistent fireball & boomerang issue as it pertained to this feature.
- This was added on this branch because it has the most consistent means of eliminating enemies, and the changes were quite small. -

royjpark commented:
> Just took a look at this and ran it on my machine with no problems. Aside from a very few instances of wonky behavior, I think the hitboxes are passable now (as far as interactions with link's swords go). The A.I. pathfinding isn't really a concern for me, personally. The biggest things are that the NPCs can damage and be damaged. With that being said though, I think the NPC hitboxes could use a slight reduction (see how Link's hitbox was reduced), because the water rooms are not currently passable.
> 
> EDIT: Upon further investigation, the Boss also has a few discrepancies:
> 1. It does not show any kind of damage (This doesn't need to be robust, but some basic feedback to show the user that the
> hit was successful is necessary)
> 2. Its movement is extremely slow and could probably use the RNG treatment that the other NPCs got.
> 
> EDIT2: Sorry for committing here, had to get this task in while I could. Should be easier to merge everything than if I took out a separate branch.

All good, will fix those asap.

bhasseman commented on CrossPlatformDesktopProject/CrossPlatformDesktopProject/NPC/Bat/Bat.cs:
```diff
@@ -9,18 +9,49 @@ class Bat : INpc
+
+            movementRNG = random.Next(1, 8);
+
+            switch(movementRNG)
```
I think you could take the switch and put it into Update, then remove the switch from all the direction classes and rely solely on the overarching Bat.cs class to determine what the next state will be. That would be a bit cleaner and prevent the large amount of repetition between the bat classes, assuming it would work how I expect it to.

royjpark replied:
I could've implemented it wrong, but I tried doing that and it made the NPCs very buggy. Might be because I use a counter within each directional class to determine when the next state is. So I tried putting the counter inside the bat class but yeah it didn't end up too hot. Might try and work on fixing this later more thoroughly though.

royjpark committed:
added damaged boss state and made movement cycle better

royjpark committed:
finished

royjpark committed:
Merge branch 'master' of https://github.com/sweeneyde/3902-zelda-game

royjpark committed:
merge w master

royjpark committed:
changed starting room back to 1
